KAFKA-9377: Refactor StreamsPartitionAssignor Repartition Count Logic

feyman2016 opened the pull request:
This PR is mainly to make StreamsPartitionAssignor more efficient in repartition count logic.

**Method to be refactored:**
```
    private void setRepartitionTopicMetadataNumberOfPartitions(final Map<String, InternalTopicConfig> repartitionTopicMetadata,
                                                               final Map<Integer, TopicsInfo> topicGroups,
                                                               final Cluster metadata)
```

What do we need to achieve?
Calc the number of repartitions for all `repartitionSourceTopics` in all the `TopicsInfo` provided. 
Rule: 
1) If a topic is one of the sink topics of a TopicInfo, use the maximum of all the TopicInfo's source topic partitions as the number of partitions.
2) If a topic is contained by more than one TopicInfos, use the maximum of all these TopicInfos' source topic partitions as the number of partitions.
To be more concrete, see the graph below, we will have(We use `tx` to refer to `NumOfPartition(TopicNode("tx") )` for simplicity):

```
     t3 = max(t1,t2)
     t4 = max(max(t1,t2), max(t6,t7))
```
In this PR, we create the `TopicNode` to represent the topic node.

The simple example of a graph built of `TopicNode` and `TopicsInfo`
```
     TopicNode("t1")      TopicNode("t2")                       TopicNode("t6")          TopicNode("t7")
                             \           /                                                            \                           /
    TopicsInfo(source = (t1,t2), sink = (t3,t4))                    TopicsInfo(source = (t6,t7),sink = (t4))
                                /           \                                                                        /
                             /                 \                                                          /
                          /                        \                                           /
                      /                                \                           /
                 /                                       \            /
     TopicNode("t3")                      TopicNode("t4")
            \
     TopicsInfo(source = (t3), sink = ())
```

How do we achieve it?
We can solve this by recursion with memoization.
1) Build the graph constructed with `TopicNode` and `TopicsInfo`, collect all the topics that need to calculate repartition count 
2) For all the topics to be calculated, find the corresponding `TopicNode`, use the maximum of all its upstream `TopicsInfo` number of partitions as its number of partitions.
3) For a given `TopicsInfo`, use the maximum of all its `sourceTopics` number of partitions as its number of partitions. So basically it's a mutual recursion.
4) For the topics that have no upstream `TopicInfo`, it should be the ending of the recursion and we can find the partition count of these topics in the parameter `metadata`.

**NB:**
1) We didn't mention the memoization part in the above algo description, but we have it in the impl for efficiency.

**Assumptions(my understanding, please correct me if it is not the case ):**
1) There should be no cycle dependency
2) All the `TopicsInfo` will have non-empty `sourceTopics`, but some `TopicsInfo` may have empty `sinkTopics`, we can deduce that the top level node in the graph must be `TopicNode`, so we can definitely terminate the recursive function.




### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

feyman2016 committed:
Merge pull request #1 from apache/trunk

Sync kafka to private fork

feyman2016 committed:
Merge pull request #2 from apache/trunk

Sync with apache kafka

feyman2016 committed:
refactor StreamsPartitionAssignor

feyman2016 committed:
update

feyman2016 committed:
update

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -147,6 +147,46 @@ public String toString() {
     }
 
+    /**
+     * TopicNode is the a topic node abstraction for graph built with TopicNode and TopicsInfo, the graph is useful
```
is a topic node

feyman2016 replied:
Fixed

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -147,6 +147,46 @@ public String toString() {
     }
 
+    /**
+     * TopicNode is the a topic node abstraction for graph built with TopicNode and TopicsInfo, the graph is useful
```
`with TopicNode and TopicsInfo` looks weird to phrase like this when we are defining `TopicNode`, maybe something like `with topic information associated with each node`.

feyman2016 replied:
Updated

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -147,6 +147,46 @@ public String toString() {
 
+    /**
+     * TopicNode is the a topic node abstraction for graph built with TopicNode and TopicsInfo, the graph is useful
+     * when in certain cases traverse is needed. For example, method setRepartitionTopicMetadataNumberOfPartitions
```
do a link: 
`method setRepartitionTopicMetadataNumberOfPartitions` -> `{@link #setRepartitionTopicMetadataNumberOfPartitions}`

feyman2016 replied:
Updated

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
-                            repartitionTopicMetadata.get(topicName).setNumberOfPartitions(numPartitions);
-                        }
-                    }
+                                                                final Cluster metadata) {
```
nit: format

feyman2016 replied:
fixed

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+        final Map<String, TopicNode> builtTopicNodes = new HashMap<>();
+        // 1.  Build a graph contains the TopicsInfo and TopicsNode
+        for (final TopicsInfo topicsInfo : topicGroups.values()) {
+            for (final String topicName : topicsInfo.repartitionSourceTopics.keySet()) {
```
Could replace with addAll:
`allRepartitionSourceTopics.addAll(topicsInfo.repartitionSourceTopics.keySet());`

feyman2016 replied:
cool, updated

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+                                                                final Cluster metadata) {
+        final Set<String> allRepartitionSourceTopics = new HashSet<>();
+        final Map<String, TopicNode> builtTopicNodes = new HashMap<>();
+        // 1.  Build a graph contains the TopicsInfo and TopicsNode
```
`a graph containing`, and correct space between `1. Build`

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+                allRepartitionSourceTopics.add(topicName);
+            }
+            for (final String sourceTopic : topicsInfo.sourceTopics) {
+                if (builtTopicNodes.containsKey(sourceTopic)) {
```
we could use `computeIfAbsent`

feyman2016 replied:
updated

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+            }
+
+            for (final String sinkTopic : topicsInfo.sinkTopics) {
+                if (builtTopicNodes.containsKey(sinkTopic)) {
```
Same here

feyman2016 replied:
fixed also

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
-        } while (numPartitionsNeeded);
+        }
+
+        // 2. Use DFS along with memoization to calc repartition number of all repartitionSourceTopics
```
memoization -> memorization ?

feyman2016 replied:
I think it should be memoization here -> https://en.wikipedia.org/wiki/Memoization

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+                                           final Map<String, InternalTopicConfig> repartitionTopicMetadata,
+                                           final Cluster metadata,
+                                           final Map<String, TopicNode> builtTopicNodes,
+                                           final Map<TopicsInfo, Integer> topicsInfoNumberOfPartitions) {
```
Do we need `topicsInfoNumberOfPartitions` as parameter? It is reset to new map for every call.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -147,6 +147,46 @@ public String toString() {
+    private static class TopicNode {
+        public final String topicName;
+        public final Set<TopicsInfo> upStreams; // upStream TopicsInfo's sinkTopics contains this
+        public final Set<TopicsInfo> downStreams; // downStreams TopicsInfo's sourceTopics contains this
```
I haven't seen this struct being used.

feyman2016 replied:
It's used to build the graph together with `TopicsInfo`

abbccdda replied:
Are you sure? I don't see this struct being used for read anywhere.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+        }
+    }
+
+    private int calcRepartitionNumForTopicInfo(final TopicsInfo topicsInfo,
```
`calcRepartitionNumForTopic` and `calcRepartitionNumForTopicInfo` are pretty hard to differentiate, better to name it more distinct.

feyman2016 replied:
Yeah, seems to be the case, but in the mutual recursive function, it seems acceptable...
Anyway, I can rename them later if think the algo is finalized

abbccdda reviewed (commented):
Thanks for the PR. Left a couple of comments. Could you also write the algorithm description in the PR description? Also although this is a refactoring, let's try to brainstorm some more advanced unit tests to ensure this logic works properly.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -513,66 +553,94 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+                                           final Map<String, InternalTopicConfig> repartitionTopicMetadata,
+                                           final Cluster metadata,
+                                           final Map<String, TopicNode> builtTopicNodes,
+                                           final Map<TopicsInfo, Integer> topicsInfoNumberOfPartitions) {
```
Does `TopicsInfo` work for a map key? Looking at its override equals, it doesn't seem to check all fields for equality.

feyman2016 replied:
Good catch! Seems problematic, I will check it later soon

feyman2016 committed:
update

feyman2016 commented:
@abbccdda Thanks for the comments, I updated the PR with few comments and added algo description. Since the algo is still under discussion, I can add the unit tests later once we have the conclusion.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -514,65 +554,81 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+            }
+        }
+
+        // 2. Use DFS along with memoization to calc repartition number of all repartitionSourceTopics
```
calc -> calculate

feyman2016 replied:
fixed

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -514,65 +554,81 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
+            } else {
+                // else calculate the max numRepartitions of its upStream TopicsInfo and set the repartitionTopicMetadata for memoization before return
+                final TopicNode topicNode = builtTopicNodes.get(topic);
+                Integer maxNumberPartitions = 0;
```
We could just use int.

feyman2016 replied:
Updated this a little bit, use Integer initial `null` to represent uncalculated

abbccdda reviewed (commented):
After looking again at the implementation, I'm convinced this should work. We could start brainstorming on the new unit tests.

feyman2016 commented:
@abbccdda Sorry for having being silent for several days, I have some local update to fix the issue that `TopicsInfo` is not appropriate for map key, but I have been caught in too many flaky integration tests locally... so haven't push them till now, will update soon

feyman2016 committed:
fix TopicsInfo non keyable issue

feyman2016 commented:
@abbccdda Updated to fix the TopicsInfo non-keyable issue.
However, this impl will not pass test: `KTableKTableForeignKeyInnerJoinMultiIntegrationTest#shouldInnerJoinMultiPartitionQueryable` and will stackoverflow because of infinite recursive call. Checked it a little bit, it is because there are cycle in the built graph:
```
TopicsInfoA -> Internal topic 22 -> TopicsInfo B -> Internal topic 30 -> TopicsInfo A
```
in the current impl, it won't fail because it will find the `numOfRepartitions` for a topics unless none of its upstream topics' `numOfRepartitions` can be found.
I'm not sure if we have the assumption that there will be no cycle in the graph built with TopicsInfo and TopicsNode, if yes, I think the test `shouldInnerJoinMultiPartitionQueryable` should be fixed, if not, I will update the algo to consider the cycle handling.

abbccdda commented:
test this please

feyman2016 commented:
```
[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.12/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:84:1: Class Fan-Out Complexity is 52 (max allowed is 50). [ClassFanOutComplexity]
```

Wired the ClassFanOutComplexity violation was not captured during my local run of checkstyle:
```
./gradlew checkstyleMain checkstyleTest spotbugsMain spotbugsTest spotbugsScoverage compileTestJava
```
Should I suppress the `ClassFanOutComplexity` check for `StreamsPartitionAssignor`?

abbccdda commented:
@feyman2016 Do you think we could have a fix for that? https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/metrics/ClassFanOutComplexityCheck.html

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -147,6 +147,71 @@ public String toString() {
+    }
+
+    // Node wrapper for TopicsInfo, which can be used together with TopicNode to build a graph to calculate
+    // numOfRepartitions, and numOfRepartitions of underlying TopicsInfo can be cached for memoization.
```
`...build a graph to calculate partition number of repartition topic, and numOfRepartitions of underlying TopicsInfo is used for memoization.`

feyman2016 replied:
fixed

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -147,6 +147,71 @@ public String toString() {
+            this.numOfRepartitions = Optional.empty();
+        }
+
+        public void setNumOfRepartitions(final int numOfRepartitions) {
```
nit: we could require non-negative value for `numOfRepartitions`

feyman2016 replied:
fixed

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -511,68 +576,86 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
      */
-    private void setRepartitionTopicMetadataNumberOfPartitions(final Map<String, InternalTopicConfig> repartitionTopicMetadata,
+    // visible for testing
+    public void setRepartitionTopicMetadataNumberOfPartitions(final Map<String, InternalTopicConfig> repartitionTopicMetadata,
```
package level access should be fine.

feyman2016 replied:
sure

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignorTest.java:
```diff
@@ -1695,6 +1698,60 @@ public void shouldThrowIfV2SubscriptionAndFutureSubscriptionIsMixed() {
+        //         topicsInfo3
+        // t1, t2... are topics are the sourceTopics/sinkTopics of these sub-topologies, numberOfPartitions of a given
+        // topic should be the maximum of its upstream topics' numberOfPartitions.
+        // For example(use NoP instead of numberOfPartitions for simplicity):
```
s/ `NoP instead of numberOfPartitions for simplicity` / `NoP = numberOfPartitions`

feyman2016 replied:
updated

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:
```diff
@@ -147,6 +147,71 @@ public String toString() {
     }
 
+    /**
+     * A TopicNode is a node that contains topic information and upstream/downstream TopicsInfo. Graph built of TopicNode and TopicsInfoNode is useful
```
Like we discussed offline, the downstream is actually not necessary.

feyman2016 replied:
yeah, absolutely, updated :)

feyman2016 committed:
Merge branch 'trunk' into KAFKA-9377

feyman2016 committed:
refactor to workaround ClassFanOutComplexity

feyman2016 committed:
update

feyman2016 commented:
@abbccdda I merged the trunk and found that `StreamsPartitionAssignor` has been updated by other PRs so it's `ClassFanOutComplexity` goes up to 49, and it will definitely exceed the upper limit 50 with my change, so I just introduced an inner class `NumOfRepartitionsCalculator` to encapsulate the new algo to workaround the `ClassFanOutComplexity` limitation.
Now we can come back to the most important issue I mentioned above, do we have the assumption that we won't have cycles in a graph built with `TopicsNode` and `TopicsInfoNode`? I will still try to find the answer for that but would be good if someone knows more details about it

abbccdda commented:
@feyman2016 Thanks for the question. I'm thinking whether it is possible to add a cycle detection logic when building this graph. In theory, Streams topology should never have cycles for sure, let me discuss with @guozhangwang offline.

feyman2016 commented:
@abbccdda Thanks! I also think it makes more sense to have no cycles with the Stream topology~
